### ISSUE TITLE
Working with multiselects components

### ISSUE THREAD

**Issue by slybridges** [erikras/redux-form#360]
:wave: @erikras 

Apologies for keeping submitting queries, but I really am enjoying `redux-form`... and hence using it a lot :). 

It now works perfectly for deep nested arrays managed directly by the librairy. But I have a case where it has to be managed outside of it. It was working prior to v3.1 of `redux-form`and now breaks.

One of my components uses `react-select` with the multiselect options (see example 2 [here](http://jedwatson.github.io/react-select/), source code [here](https://github.com/JedWatson/react-select/blob/master/examples/src/components/Multiselect.js)). In this mode, `react-select` component is taking care of managing the array of selected values.

I tried to integrate it with `redux-form` with it in 2 ways
1. I tried to declare my field an non-array and thinking that would trick `redux-form` in just taking the initial value I passed to it, even if this value is an array. That didn't work, `redux-form`went to each element of the array and initialized them instead.
2. I declared the field as an array, initialized it with an array and `redux form` was happy. Then, as `react-select` manages that array, I tried to pass the whole array field to the component. But it complained that it didn't find the `onBlur`and `onChange` props. This is because thoses functions are only attached to the individual elements and not the arrays or objects in between.

Any idea how to handle this case ? Shouldn't `redux-form` not try to be oversmart and not go through individual value keys if the field is declared as a non-array, non-object value field ?

I wrote the two tests below to describe the 2 situations I tried. 
Unless there are some mysterious ways unknown to me, I think we should make at least one of the two pass.

```javascript
  it('should initialize a non-array field with an array value and let it read it back', () => {
    const store = makeStore();
    const form = 'testForm';
    const Decorated = reduxForm({
      form,
      fields: ['children'],
      initialValues: {
        children: [1, 2]
      }
    })(Form);
    const dom = TestUtils.renderIntoDocument(
      <Provider store={store}>
        <Decorated/>
      </Provider>
    );
    const stub = TestUtils.findRenderedComponentWithType(dom, Form);

    expectField({
      field: stub.props.fields.children,
      name: 'children',
      value: [1, 2],
      valid: true,
      dirty: false,
      error: undefined,
      touched: false,
      visited: false
    });
  });

// FAILS ===>  Error: Expected undefined to be [ 1, 2 ]

  it('should attach form handling fields at array level', () => {
    const store = makeStore();
    const form = 'testForm';
    const Decorated = reduxForm({
      form,
      fields: ['children[]'],
      initialValues: {
        children: [1, 2]
      }
    })(Form);
    const dom = TestUtils.renderIntoDocument(
      <Provider store={store}>
        <Decorated/>
      </Provider>
    );
    const stub = TestUtils.findRenderedComponentWithType(dom, Form);

    expectField({
      field: stub.props.fields.children,
      name: 'children',
      value: [1, 2],
      valid: true,
      dirty: false,
      error: undefined,
      touched: false,
      visited: false
    });
  });

// FAILS ===> Error: Expected undefined to be 'children'
```

... sorry for the long read !

**Comment by erikras** [comment]
Very interesting point, and fantastic that you provided a test case. :trophy: issue submission!

I'm using that same multiselect library in one of my own projects, but I'm not up to v3.1 yet. Will investigate...

**Comment by motoedie** [comment]
I've got the same problem. I had to revert back to 3.1.0

**Comment by erikras** [comment]
Of your two tests, the former should pass.

**Comment by slybridges** [comment]
agreed, the first one is the most logical way you want to use it

**Comment by erikras** [comment]
This is pretty complicated, actually, since the error is happening in the reducer when it gets the `INITIALIZE` action, but at that point, it doesn't have the fields list, just the data object, so it doesn't/can't know that you don't have the `[]` on the field name.

**Comment by slybridges** [comment]
can't we wait to have the fields before `INITIALISE`ing and passe them along as attribute of the data object ?

**Comment by erikras** [comment]
We _have_ the `fields` array at the time that we are initializing using the `initialValues` prop, but it also allows someone to dispatch the `initialize()` action themselves, so we can't require that it also have the `fields`. I guess it will just be smart about arrays when it has the `fields` and not when it doesn't...

**Comment by trashgenerator** [comment]
I found a workaround for now. You can create a plugin to fix INITIALIZE action manualy. I'm not sure that everything is correct but it works for me.

if you have that form:
```
reduxForm({
    form: 'YourFormName',
    fields: ['yourMultiselectField', 'someOtherField'],
    initialValues: {
        'yourMultiselectField': [1,2,3],
        'someOtherField': 'some value',
    },
})(YourComponent);
```

Your reducers:
```
export default combineReducers({
  form: formReducer.plugin({
    YourFormName: (state, action) => {
      switch(action.type) {
        case INITIALIZE:
          return {
            ...state,
            yourMultiselectField: {
              value: action.data.yourMultiselectField,
              initial: action.data.yourMultiselectField,
            },
          };
        default:
          return state;
      }
    }
  }),
```

**Comment by erikras** [comment]
This should be fixed in [`v4.0.0`](https://github.com/erikras/redux-form/releases/tag/v4.0.0). Come and get it!
